toolchain/check/testdata/facet/period_self.carbon

```diff
@@ -0,0 +1,517 @@
+  // CHECK:STDERR:   let a: U = U.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR:
+  let a: U = U.G();
+}
+
+// --- fail_todo_return_of_type_period_self_assoc_const_has_type_u.carbon
+library "[[@TEST_NAME]]";
+
+interface I {
+  let X:! type;
+  fn G() -> X;
+}
+
+fn F(U:! I where .X = .Self) {
+  // CHECK:STDERR: fail_todo_return_of_type_period_self_assoc_const_has_type_u.carbon:[[@LINE+7]]:14: error: cannot implicitly convert expression of type `.Self` to `U` [ConversionFailure]
+  // CHECK:STDERR:   let a: U = U.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR: fail_todo_return_of_type_period_self_assoc_const_has_type_u.carbon:[[@LINE+4]]:14: note: type `.Self` does not implement interface `Core.ImplicitAs(U)` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   let a: U = U.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR:
+  let a: U = U.G();
+}
+
+// --- fail_todo_nested_period_self.carbon
+library "[[@TEST_NAME]]";
+
+interface I(T:! type) {
+  let A:! type;
+  let B:! type;
+  fn G() -> T;
+}
+
+// Both `.Self` refer to `T`. The first because it's the interface for the
+// binding. The second because it refers to the top level facet type which is
+// constraining the binding.
+fn F(T:! I(.Self) where .A = ((I(.Self) where .B = {}) where .A = {}) and .B = {}, U:! T.A) {
+  // T.G() has type T.
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+7]]:14: error: cannot implicitly convert expression of type `.Self` to `T` [ConversionFailure]
+  // CHECK:STDERR:   let t: T = T.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+4]]:14: note: type `.Self` does not implement interface `Core.ImplicitAs(T)` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   let t: T = T.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR:
+  let t: T = T.G();
+    // U.G() has type T.
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+7]]:14: error: cannot implicitly convert expression of type `.Self` to `T` [ConversionFailure]
+  // CHECK:STDERR:   let u: T = U.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+4]]:14: note: type `.Self` does not implement interface `Core.ImplicitAs(T)` [MissingImplInMemberAccessNote]
+  // CHECK:STDERR:   let u: T = U.G();
+  // CHECK:STDERR:              ^~~~~
+  // CHECK:STDERR:
+  let u: T = U.G();
+
+  // Shows both `I(.Self)` are `I(T)`.
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+4]]:9: error: found cycle in facet type constraint for `.(I(T).A)` [FacetTypeConstraintCycle]
+  // CHECK:STDERR:   T as (I(T) where .A = (I(T) where .A = {} and .B = {}));
+  // CHECK:STDERR:         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  T as (I(T) where .A = (I(T) where .A = {} and .B = {}));
```

Comment:
I don't think this should actually be a cycle, but it's still being reported as one when I get .Self working more, so pointing this out to double check.